Add Unified Resource watcher

avatus opened the pull request:
In support of https://github.com/gravitational/teleport/pull/28162
Closes https://github.com/gravitational/teleport/issues/28322
Updates #27462

This pr creates the unified resource watcher, an in-memory store of resources relevant to the web UI. It uses a similar structure to our memory backend, but stores the resource itself rather than a byte representation.

The bulk of the changes are in the new lib/services/unified_resource.go

avatus committed:
Adds a UnifiedResource watcher to the auth server. This watcher
will watch all the types that are displayed in the web UI and store
them in-memory to allow us to search/filter/query and get multiple
kinds returned at the same time.
[RFD](https://github.com/gravitational/teleport/pull/28162)

zmb3 commented on lib/services/unified_resource_test.go:
```diff
@@ -0,0 +1,102 @@
+	expectedRes := []types.ResourceWithLabels{node, db, app, win}
+	assert.Eventually(t, func() bool {
+		res, err = w.GetUnifiedResources(ctx)
+		require.NoError(t, err)
```
We can't use `require` in an `Eventually`, because the `func() bool` runs in a backround goroutine, and `require` assertions fail the test, which can only be done from the test's main goroutine.

zmb3 commented on lib/services/unified_resource_test.go:
```diff
@@ -0,0 +1,102 @@
+		res, err = w.GetUnifiedResources(ctx)
+		require.NoError(t, err)
+		return len(res) == len(expectedRes)
+	}, time.Second, time.Millisecond, "Timed out waiting for unified resources")
```
1 second is likely not enough time to avoid flakiness in CI, I would bump this up to 3-5s.

avatus commented:
Ok, taking this out of draft mode and marking as ready for review. Wasn't sure who else to add on as reviewers, feel free to add anyone else who might be interested.

avatus commented:
Added the grpc endpoint PR [here](https://github.com/gravitational/teleport/pull/29661)

avatus commented on lib/services/watcher.go:
```diff
@@ -741,6 +746,10 @@ func lockMapValues(lockMap map[string]types.Lock) []types.Lock {
 	return locks
 }
 
+type DatabaseServersGetter interface {
+	GetDatabaseServers(context.Context, string, ...MarshalOption) ([]types.DatabaseServer, error)
+}
```
this is an awesome resource, thank you! i will move them

fspmarshall commented on lib/services/unified_resource.go:
```diff
@@ -0,0 +1,675 @@
+			return
+		}
+		c.tree.Delete(item)
+	default:
+		// skip unsupported record
+	}
```
Likely indicates a bug and should be logged.

fspmarshall commented on lib/services/unified_resource.go:
```diff
@@ -0,0 +1,675 @@
+	Clock clockwork.Clock
+	// Component is a logging component
+	Component string
+	// EventsOff turns off events generation
+	EventsOff bool
```
This config option is never used.

fspmarshall commented on lib/services/unified_resource.go:
```diff
@@ -0,0 +1,675 @@
+func (u *UnifiedResourceWatcher) GetUnifiedResources(ctx context.Context) ([]types.ResourceWithLabels, error) {
+	var resources []types.ResourceWithLabels
+
+	// if the watcher is not initialized or stale, instead of returning nothing, return upstream nodes
+	if !u.IsInitialized() || u.stale {
+		nodes, err := u.NodesGetter.GetNodes(ctx, apidefaults.Namespace)
+		if err != nil {
+			return nil, trace.Wrap(err, "getting nodes while unified resource cache is uninitialized or stale")
+		}
+
+		for _, node := range nodes {
+			resources = append(resources, node)
+		}
+		return resources, nil
+	}
```
We shouldn't omit non-node resources if it's uninitialized/stale.  Keeping consistent behavior across initialized and uninitialized state is very important (should also have test coverage to ensure that initialized and uninitialized state are indistinguishable to the caller).  Additionally, we probably don't want to be deserializing a fresh copy per incoming request.

I realize this is a non-trivial ask.  I'd suggest handling this by maintaining a separate `unifiedResourceCache` instance that isn't live (i.e. has no watcher and can be initialized immediately as-needed), and initializing it lazily as-needed, re-creating it on some TTL.  Basically, have a snapshot of a cache that you can create once and serve unlimited requests from while the primary cache is unhealthy, then reset it every 10-15s to ensure it doesn't ever get too stale.

fspmarshall replied:
@zmb3 The primary cache uses a replication strategy.  In order to do that, it cannot load its initial state until after the watcher reports healthy, otherwise the cache becomes invalid and may miss put/delete events.

Falling back to a tll-based caching strategy if the primary cache is unhealthy is how `cache.Cache` works, and while its a bit heavy-weight/annoying, its proven to be fairly resilient even in the face of serious bugs/network instability.  The only other alternatives I'm aware of are:

1. Load everything for every request (same amount of work as creating a new cache per request, strictly-worse than ttl-based fallback if 2 or more requests come in).
2. Return an error if primary cache is uninitialized (flaky, poor resilience to bugs/instability).
3. Block, waiting for cache to become healthy (totally fine as long as the failure is fairly transient, but worst possible outcome if error is persistent or semi-persistent).

Note: this cache reads from `cache.Cache`, not the real backend.  Loading and building an extra cache state isn't cheap, but it's cheap enough that it's probably better than returning en error and/or blocking.

fspmarshall commented on lib/services/unified_resource.go:
```diff
@@ -0,0 +1,675 @@
+	return u.unifiedResourceCollector.Close()
+}
+
+// GetUnifiedResources returns a list of all resources stored in the current unifiedResourceCollector tree
+func (u *UnifiedResourceWatcher) GetUnifiedResources(ctx context.Context) ([]types.ResourceWithLabels, error) {
```
The primary motivation of creating a unified resource watcher was to ensure that we _didn't_ need to load the entire resource set into memory simultaneously for each request, and that we'd instead be able to paginate and pre-filter.  Loading the entire resource set into memory per-request is a recipe for OOM errors (not hypothetical, we see this kind of thing happen in real teleport deployments).  This should be a paginated API.  Nothing wrong with having a "get all" method for tests, but it should be private to ensure folks don't accidentally use it to handle actual requests.

fspmarshall commented on lib/services/unified_resource.go:
```diff
@@ -0,0 +1,675 @@
+	if err != nil {
+		return nil, trace.Wrap(err, "getting unified resource range")
+	}
+	for _, item := range result.Items {
+		resources = append(resources, item.Value)
+	}
```
This is a risky default.  Any resource returned to the caller should be deep copied to prevent concurrent memory modification. If we discover a code path where we are 100% certain no modification occurs we could maybe add some optimizations down the line, but the default should be to clone on get.

Additionally, we should only be returning/copying those resources that match our query/filter (see the node watcher for an example of how to do this, a simple `match fn(Resource) bool` is probably the easiest).

zmb3 commented on lib/services/unified_resource.go:
```diff
@@ -0,0 +1,675 @@
+}
+
+// GetUnifiedResources returns a list of all resources stored in the current unifiedResourceCollector tree
+func (u *UnifiedResourceWatcher) GetUnifiedResources(ctx context.Context) ([]types.ResourceWithLabels, error) {
+	var resources []types.ResourceWithLabels
+
+	// if the watcher is not initialized or stale, instead of returning nothing, return upstream nodes
+	if !u.IsInitialized() || u.stale {
+		nodes, err := u.NodesGetter.GetNodes(ctx, apidefaults.Namespace)
+		if err != nil {
+			return nil, trace.Wrap(err, "getting nodes while unified resource cache is uninitialized or stale")
+		}
+
+		for _, node := range nodes {
+			resources = append(resources, node)
+		}
+		return resources, nil
+	}
+
+	result, err := u.current.GetRange(ctx, backend.Key(prefix), backend.RangeEnd(backend.Key(prefix)), backend.NoLimit)
+	if err != nil {
+		return nil, trace.Wrap(err, "getting unified resource range")
+	}
+	for _, item := range result.Items {
+		resources = append(resources, item.Value)
+	}
+
+	return resources, nil
+}
```
+1

avatus commented on lib/services/unified_resource.go:
```diff
@@ -0,0 +1,675 @@
+
+// Put puts value into backend (creates if it does not
+// exist, updates it otherwise)
+func (c *unifiedResourceCache) Put(ctx context.Context, i Item) error {
```
Definitely. I ended up making those private and removing `CompareAndSwap` and `Update` because they weren't used anyway.

avatus commented:
Updated a lot of the little things in the review feedback but will tackle moving `GetUnifiedResources` away from the watcher and the Deep Copy thing in the morning. 

After running `loadtest` with 30k nodes for 20 minutes, I found that the unified resource watcher would overflow. Upping it's QueueSize to `8192` (to match the auth/proxy queue sizes) solved the problem and ran without a hitch for 30 minutes or so.

Also, manual testing on the web ui showed no signs of increased latency compared to previous list resources (with the 30k nodes still slammin).

Going to continue to run tests but the initial findings are good (i think!)

avatus commented:
Ok, updated with all the feedback besides a few things below. the rest i believe is ready for another round of feedback. 
- didnt update the test to make it simpler yet, but it does what we want still
- this comment from forrest, https://github.com/gravitational/teleport/pull/28296#discussion_r1276540131, we are going to skip. currently, the web UI always "gets all" for every resource as we need to use the full list for sorting/checking access/and filtering. we plan to improve this in the future
- still need to figure this out https://github.com/gravitational/teleport/pull/28296#discussion_r1276531679 

the significant changes are, the cache is now what Gets the unified resources, and i've added a new interface for copying since for some reason, a few of our resources have different clone/copy implementations. I've added the missing few for desktops and SAML apps. I don't think its the best place to put it in this PR but, for sake of time its in here (and still at least makes some sense as its related).

avatus commented on lib/services/unified_resource.go:
```diff
@@ -0,0 +1,515 @@
+		ResourceGetter:  cfg.ResourceGetter,
+	}
+
+	err := newWatcher(ctx, m, cfg.ResourceWatcherConfig)
```
the watcher requires the cache to be created _before_ we init it since the cache is an argument for the watcher (its the collector used in `newResourceWatcher`). I might be mistaken. 

Other examples I've seen have the watcher init first, but the collector inside of their newwatcher method. So I think one has to come before the other which ever way we do it. I chose this way so that I'd be returning a cache to set on the authServer instead of the watcher like before.

avatus commented on lib/services/unified_resource_test.go:
```diff
@@ -0,0 +1,213 @@
+
+	ctx := context.Background()
+	conf, bk := setupConfig(t, ctx)
+	authServer, err := auth.Init(ctx, conf)
```
I need the authserver as the `ResourceGetter` because the `presence` doesn't have access to all the getters I need (like windowsdesktopgetter). Am I misunderstanding the intent? (the test was updated a bit since this comment)

avatus commented on lib/services/unified_resource.go:
```diff
@@ -0,0 +1,515 @@
+}
+
+// getAndUpdateNodes will get nodes and update the current tree with each Node
+func (u *UnifiedResourceCache) getAndUpdateNodes(ctx context.Context) error {
```
Done

fspmarshall commented on lib/services/unified_resource.go:
```diff
@@ -0,0 +1,464 @@
+		return trace.Wrap(err)
+	}
+
+	c.stale = false
```
It isn't safe to read or write `c.stale` when not holding the lock.

fspmarshall commented on lib/services/unified_resource.go:
```diff
@@ -0,0 +1,464 @@
+	return backend.Key(prefix, resource.GetName(), resource.GetKind())
+}
+
+func (c *UnifiedResourceCache) getResourcesAndUpdateCurrent(ctx context.Context) error {
+	err := c.getAndUpdateNodes(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
```
We need to clear the previous state (i.e. delete all existing items) before inserting the new state, otherwise the cache may end up containing invalid items.  Consider the following scenario:

- Cache initializes and becomes healthy with items `A` and `B`.
- The watcher closes due to error (e.g. network interruption).
- Item `A` is deleted from the backend at and then item `C` is inserted into the backend.  The cache does not observe the events for these operations since its watcher is unhealthy.
- Cache re-establishes a healthy watcher and loads items `B` and `C` from the backend.

Without clearing first, the cache now contains items `A`, `B`, `C`, which is categorically invalid since `A` was deleted before `C` was inserted (i.e. its not just presenting an outdated/stale state... its presenting a combination of items that were never a valid state).

fspmarshall commented on lib/services/unified_resource.go:
```diff
@@ -0,0 +1,464 @@
+			return nil, nil
+		}
+
+		c.tree = fallbackCache.tree
```
Having the ttl-based fallback tree take up the same slot as the primary tree complicates the concurrency model, since this means it isn't safe to read from the fallback tree while the primary tree is being initialized (since you might end up reading from a partially-initialized state). This isn't that crazy of a thing to workaround, the main thing is to make sure that your initialization/reset logic holds the lock continuously, so that no reads happen between clearing the ttl-tree state and completion of all inserts of new values.

Not a requirement, but I'm a fan of creating helper methods that handle deciding which underlying state to access and abstract over the different locking requirements of the two access patterns.  We did a more complex version of this in `lib/cache` with the "read guard", but I think a simpler model of just passing a closure into a helper function would work pretty well here:

```golang
// read applies the supplied closure to either the primary tree or the ttl-based fallback tree depending on
// wether or not the cache is currently healthy.  locking is handled internally and the passed-in tree should
// not be accessed after the closure completes.
func (c *UnifiedResourceCache) read(fn func(tree *btree.BTreeG[*item])) error {
    c.mu.Lock()
    
    if !c.stale {
        fn(c.tree)
        c.mu.Unlock()
        return nil
    }
    
    c.mu.Unlock()
    ttlTree, err := utils.FnCacheGet(ctx, c.cache, "unified_resources", func(ctx context.Context) (*btree.BTreeG[*item], error) {
        // ...
        return fallbackCache.tree
    })
    c.mu.Lock()

    if !c.stale {
        // primary became healthy while we were waiting
        fn(c.tree)
        c.mu.Unlock()
        return nil
    }
    c.mu.Unlock()
    
    if err != nil {
        // ttl-tree setup failed
        return trace.Wrap(err)
    }
    
    fn(ttlTree)
    return nil
}


func (c *UnifiedResourceCache) example() error {
    err := c.read(func(tree *btree.BTreeG[*item]) {
        // do stuff with the tree here while the lock is held
    })
    return trace.Wrap(err)
}
```

avatus commented on lib/services/unified_resource.go:
```diff
@@ -0,0 +1,475 @@
+	}
+}
+
+func (c *UnifiedResourceCache) getResourcesAndUpdateCurrent(ctx context.Context) error {
```
I did the one big insert but the function seems a bit ugly because I have to append to my `newResources` slice every time i run a new getter. Is there a cleaner way to do it? 

As of now, the lock should only be acquired three times. one to `clear`, one for the big `putResources` and once for updating `c.stale`. 

Also, as spoke of in DMs but putting here for posterity, this will prevent "halfstates" where say we were able to get and update nodes, but not the other resources. we wouldn't want a tree only sorta updating (even if it does stay stale).

rosstimothy reviewed (approved):
LGTM 🎉

fspmarshall commented on lib/services/unified_resource.go:
```diff
@@ -0,0 +1,514 @@
+		return trace.Wrap(err)
+	}
+
+	return c.read(ctx, func(tree *btree.BTreeG[*item]) error {
+		tree.Clear(false)
+		putResources[types.Server](tree, newNodes)
+		putResources[types.DatabaseServer](tree, newDbs)
+		putResources[types.KubeServer](tree, newKubes)
+		putResources[types.AppServer](tree, newApps)
+		putResources[types.SAMLIdPServiceProvider](tree, newSAMLApps)
+		putResources[types.WindowsDesktop](tree, newDesktops)
+
+		c.stale = false
+		c.defineCollectorAsInitialized()
+		return nil
+	})
```
This shouldn't be using `read`.  The `read` function is a helper specifically for falling back to the TTL-cache if the primary is stale/unhealthy.  Initialization is supposed to always insert into the primary `c.tree`.  I'd move this block out of `read` and just hold the lock normally here instead.  As currently implemented, the initial state will end up getting written into the ttl-cache's tree, and basically get 'lost' since we mark the primary as healthy right after insertion.

It looks like current test coverage only starts creating resources after the cache has become healthy, meaning that the test coverage effectively isn't checking if this initialization logic works (since the initial state is always an empty tree), its just checking if event handling works (which is also important).  I'd suggest adding some test coverage that ensures that the cache initialize correctly in the presence of existing resources.

fspmarshall commented on lib/services/unified_resource.go:
```diff
@@ -0,0 +1,514 @@
+// put stores the value into backend (creates if it does not
+// exist, updates it otherwise)
+func (c *UnifiedResourceCache) put(ctx context.Context, i item) error {
+	if len(i.Key) == 0 {
+		return trace.BadParameter("missing parameter key")
+	}
+	return c.read(ctx, func(tree *btree.BTreeG[*item]) error {
+		tree.ReplaceOrInsert(&i)
+		return nil
+	})
```
Insertion is always supposed to hit the primary tree.

```suggestion
	if len(i.Key) == 0 {
		return trace.BadParameter("missing parameter key")
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	c.tree.ReplaceOrInsert(&i)
	return nil
```

fspmarshall reviewed (approved):
LGTM once insertions are tweaked to always hit the primary tree (see comments).

avatus committed:
Add ListUnifiedResources gRPC and web endpoints (#29661)

Adds the `ListUnifiedResources` grpc endpoint that returns paginated unified resources and exposes it in the web apiserver

r0mant reviewed (approved):
Bot.

zmb3 commented:
/excludeflake TestClusterNodesGet TestClusterAppsGet TestClusterDesktopsGet TestCreateDatabase TestUpdateDatabase_NonErrors

zmb3 commented:
/excludeflake TestClusterDatabasesGet_NoRole

avatus committed:
Fix test

zmb3 commented:
/excludeflake *
